host/crypto/bcrypt/pem.c

```diff
@@ -0,0 +1,341 @@
+        OE_RAISE(OE_INVALID_PARAMETER);
+
+    success = CryptBinaryToStringA(
+        der_data,
+        der_data_size,
+        CRYPT_STRING_BASE64 | CRYPT_STRING_NOCR,
+        NULL,
+        &pem_local_size);
+
+    /* Should return true and set pem_local_size when given a null buffer.
+     * Resulting pem_local_size includes null terminator. */
+    if (!success)
+        OE_RAISE_MSG(
+            OE_CRYPTO_ERROR,
+            "CryptBinaryToStringA failed (err=%#x)\n",
+            GetLastError());
+
+    /* Need to allocate and write the PEM header/footer manually because
+     * BCrypt only supports the cert/CRL/CSR headers.
+     * The size also accounts for LF characters at the end of each line. */
+    pem_info = &_PEM_HEADERS[(DWORD)pem_type];
+    assert(pem_info->type == pem_type);
+    assert(pem_info->begin_label_length < OE_PEM_MAX_LEN);
+    assert(pem_info->end_label_length < OE_PEM_MAX_LEN);
+
+    {
+        /* Max pem_headers_size is (2 * OE_PEM_MAX_LEN + 2) < MAXDWORD */
+        DWORD pem_headers_size = (DWORD)(
+            pem_info->begin_label_length + pem_info->end_label_length + 2);
+        OE_CHECK(
+            oe_safe_add_u32(pem_local_size, pem_headers_size, &pem_local_size));
+
+        pem_local = (uint8_t*)malloc(pem_local_size);
+        if (pem_local == NULL)
+            OE_RAISE(OE_OUT_OF_MEMORY);
+    }
+
+    {
+        /* Write the begin public key header */
+        uint8_t* pos = pem_local;
+        DWORD size_left = pem_local_size;
+
+        OE_CHECK(oe_memcpy_s(
+            pos,
+            size_left,
+            pem_info->begin_label,
+            pem_info->begin_label_length));
+        pos += pem_info->begin_label_length;
+        size_left -= (DWORD)pem_info->begin_label_length;
+
+        /* Write a LF character */
+        *pos++ = 0x0A;
+        size_left--;
+
+        /* Write the encoded key data */
+        DWORD written_size = size_left;
+        success = CryptBinaryToStringA(
+            der_data,
+            der_data_size,
+            CRYPT_STRING_BASE64 | CRYPT_STRING_NOCR,
+            pos,
+            &written_size);
+
+        if (!success)
+            OE_RAISE_MSG(
+                OE_CRYPTO_ERROR,
+                "CryptBinaryToStringA failed (err=%#x)\n",
+                GetLastError());
+
+        /* Assert null-terminator at end of string.
+         * This method overwrites it to extend the string for the end footer */
+        pos += written_size;
+        assert(*pos == '\0');
+        size_left -= written_size;
+
+        /* Write the end public key footer */
+        OE_CHECK(oe_memcpy_s(
+            pos, size_left, pem_info->end_label, pem_info->end_label_length));
+        pos += pem_info->end_label_length;
+        size_left -= (DWORD)pem_info->end_label_length;
+
+        /* Add LF and null-terminator */
+        OE_CHECK(size_left == 2 ? OE_OK : OE_UNEXPECTED);
+        *pos++ = 0x0A;
```

Comment:
Perhaps add a define for the constants 0x0A and 0x00?

Reply:
Replaced with character literals '\n' and '\0' as used elsewhere in the code.